// plywood-mysql-requester/build/mySqlRequester.js
"use strict";
var mysql = require("mysql");
var Q = require('q');
function basicLocator(host) {
    var hostnamePort = host.split(':');
    var hostname;
    var port;
    if (hostnamePort.length > 1) {
        hostname = hostnamePort[0];
        port = Number(hostnamePort[1]);
    }
    else {
        hostname = hostnamePort[0];
        port = 3306;
    }
    return function () {
        return Q({
            hostname: hostname,
            port: port
        });
    };
}
function mySqlRequesterFactory(parameters) {
    var locator = parameters.locator;
    if (!locator) {
        var host = parameters.host;
        if (!host)
            throw new Error("must have a `host` or a `locator`");
        locator = basicLocator(host);
    }
    var user = parameters.user;
    var password = parameters.password;
    var database = parameters.database;
    return function (request) {
        var query = request.query;
        return locator()
            .then(function (location) {
            var connection = mysql.createConnection({
                host: location.hostname,
                port: location.port || 3306,
                user: user,
                password: password,
                database: database,
                charset: 'UTF8_BIN',
                timezone: '+00:00'
            });
            connection.connect();
            var deferred = (Q.defer());
            connection.query(query, function (err, data) {
                if (err) {
                    deferred.reject(err);
                }
                else {
                    deferred.resolve(data);
                }
            });
            connection.end();
            return deferred.promise;
        });
    };
}
exports.mySqlRequesterFactory = mySqlRequesterFactory;
