// immutable-class/build/base-immutable/base-immutable.js
"use strict";
var utils_1 = require('../utils/utils');
var equality_1 = require('../equality/equality');
function firstUp(name) {
    return name[0].toUpperCase() + name.substr(1);
}
var BaseImmutable = (function () {
    function BaseImmutable(value) {
        var properties = this.ownProperties();
        for (var _i = 0, properties_1 = properties; _i < properties_1.length; _i++) {
            var property = properties_1[_i];
            var propertyName = property.name;
            var pv = value[propertyName];
            if (pv == null) {
                if (!property.hasOwnProperty('defaultValue')) {
                    throw new Error(this.constructor.name + "." + propertyName + " must be defined");
                }
            }
            else {
                var possibleValues = property.possibleValues;
                if (possibleValues && possibleValues.indexOf(pv) === -1) {
                    throw new Error(this.constructor.name + "." + propertyName + " can not have value '" + pv + "' must be one of [" + possibleValues.join(', ') + "]");
                }
                if (property.isDate) {
                    if (isNaN(pv)) {
                        throw new Error(this.constructor.name + "." + propertyName + " must be a Date");
                    }
                }
                var validate = property.validate;
                if (validate) {
                    var validators = Array.isArray(validate) ? validate : [validate];
                    try {
                        for (var _a = 0, validators_1 = validators; _a < validators_1.length; _a++) {
                            var validator = validators_1[_a];
                            validator(pv);
                        }
                    }
                    catch (e) {
                        throw new Error(this.constructor.name + "." + propertyName + " " + e.message);
                    }
                }
            }
            this[propertyName] = pv;
        }
    }
    BaseImmutable.jsToValue = function (properties, js) {
        var value = {};
        for (var _i = 0, properties_2 = properties; _i < properties_2.length; _i++) {
            var property = properties_2[_i];
            var propertyName = property.name;
            var pv = js[propertyName];
            if (pv != null) {
                if (property.isDate) {
                    pv = new Date(pv);
                }
                else if (property.immutableClass) {
                    pv = property.immutableClass.fromJS(pv);
                }
                else if (property.immutableClassArray) {
                    if (!Array.isArray(pv))
                        throw new Error("expected " + propertyName + " to be an array");
                    var propertyImmutableClassArray = property.immutableClassArray;
                    pv = pv.map(function (v) { return propertyImmutableClassArray.fromJS(v); });
                }
            }
            value[propertyName] = pv;
        }
        return value;
    };
    BaseImmutable.finalize = function (ClassFn) {
        var proto = ClassFn.prototype;
        ClassFn.PROPERTIES.forEach(function (property) {
            var propertyName = property.name;
            var upped = firstUp(property.name);
            var getUpped = 'get' + upped;
            var changeUpped = 'change' + upped;
            proto[getUpped] = proto[getUpped] || function () {
                return this.get(propertyName);
            };
            proto[changeUpped] = proto[changeUpped] || function (newValue) {
                return this.change(propertyName, newValue);
            };
        });
    };
    BaseImmutable.prototype.ownProperties = function () {
        return this.constructor.PROPERTIES;
    };
    BaseImmutable.prototype.findOwnProperty = function (propName) {
        var properties = this.ownProperties();
        return properties.filter(function (p) { return p.name === propName; })[0] || null;
    };
    BaseImmutable.prototype.valueOf = function () {
        var value = {};
        var properties = this.ownProperties();
        for (var _i = 0, properties_3 = properties; _i < properties_3.length; _i++) {
            var property = properties_3[_i];
            var propertyName = property.name;
            value[propertyName] = this[propertyName];
        }
        return value;
    };
    BaseImmutable.prototype.toJS = function () {
        var js = {};
        var properties = this.ownProperties();
        for (var _i = 0, properties_4 = properties; _i < properties_4.length; _i++) {
            var property = properties_4[_i];
            var propertyName = property.name;
            var pv = this[propertyName];
            if (pv != null) {
                if (property.immutableClass) {
                    pv = pv.toJS();
                }
                else if (property.immutableClassArray) {
                    pv = pv.map(function (v) { return v.toJS(); });
                }
                js[propertyName] = pv;
            }
        }
        return js;
    };
    BaseImmutable.prototype.toJSON = function () {
        return this.toJS();
    };
    BaseImmutable.prototype.toString = function () {
        var name = this.name;
        var extra = name === 'string' ? ": " + name : '';
        return "[ImmutableClass" + extra + "]";
    };
    BaseImmutable.prototype.equals = function (other) {
        if (!other)
            return false;
        if (this === other)
            return true;
        if (!utils_1.isInstanceOf(other, this.constructor))
            return false;
        var properties = this.ownProperties();
        for (var _i = 0, properties_5 = properties; _i < properties_5.length; _i++) {
            var property = properties_5[_i];
            var propertyName = property.name;
            var equal = property.equal || equality_1.generalEqual;
            if (!equal(this[propertyName], other[propertyName]))
                return false;
        }
        return true;
    };
    BaseImmutable.prototype.get = function (propName) {
        var properties = this.ownProperties();
        for (var _i = 0, properties_6 = properties; _i < properties_6.length; _i++) {
            var property = properties_6[_i];
            if (property.name === propName) {
                var pv = this[propName];
                return pv != null ? pv : property.defaultValue;
            }
        }
        throw new Error("can not find prop " + propName);
    };
    BaseImmutable.prototype.change = function (propName, newValue) {
        var value = this.valueOf();
        var property = this.findOwnProperty(propName);
        if (!property) {
            throw new Error("Unknown property: " + propName);
        }
        value[propName] = newValue;
        return new this.constructor(value);
    };
    BaseImmutable.ensure = {
        number: function (n) {
            if (isNaN(n) || typeof n !== 'number')
                throw new Error("must be a number");
        }
    };
    return BaseImmutable;
}());
exports.BaseImmutable = BaseImmutable;
